stan/math/prim/prob/wiener_full_lpdf.hpp

```diff
@@ -0,0 +1,740 @@
+ * \c -v and \c (1-w), respectively.
+ *
+ * \c sv, \c sw, \c st0, and \c t0 can be set to zero, indicating no inter-trial
+ * variability in \f$v\f$, no inter-trial variability in \f$w\f$, no
+ * inter-trial variability in \f$t_0\f$, and no non-decision time, respectively.
+ * If \c t0 is zero, \c st0 must be zero as well. For example, when no
+ inter-trial
+ * variability for the relative starting point is needed one can write something
+ like:
+ @code
+ target += wiener_full_lpdf(y, a, v, w, t0, sv, 0, st0)
+ @endcode
+ * If no inter-trial variability is needed at all one can write something like:
+ @code
+ target += wiener_full_lpdf(y, a, v, w, t0, 0, 0, 0)
+ @endcode
+ * If for some reason no non-decision time is assumed one can write something
+ like:
+ @code
+ target += wiener_full_lpdf(y, a, v, w, 0, sv, sw, 0)
+ @endcode
+ *
+ * To also control the precision in the estimation of the partial derivatives,
+ * call the function \c wiener_full_prec_lpdf(), analogously:
+ @code
+ target += wiener_full__prec_lpdf(y, a, v, w, t0, sv, sw, st0, precision);
+ @endcode
+ *
+ *
+ * **References**
+ * - Blurton, S. P., Kesselmeier, M., & Gondan, M. (2017). The first-passage
+ * time distribution for the diffusion model with variable drift.
+ * *Journal of Mathematical Psychology, 76*, 7–12.
+ * https://doi.org/10.1016/j.jmp.2016.11.003
+ * - Foster, K., & Singmann, H. (2021). Another Approximation of the
+ First-Passage
+ * Time Densities for the Ratcliff Diffusion Decision Model.
+ * *arXiv preprint arXiv:2104.01902*
+ * - Gondan, M., Blurton, S. P., & Kesselmeier, M. (2014). Even faster and even
+ * more accurate first-passage time densities and distributions for the Wiener
+ * diffusion model. *Journal of Mathematical Psychology, 60*, 20–22.
+ * https://doi.org/10.1016/j.jmp.2014.05.002
+ * - Hartmann, R., & Klauer, K. C. (2021). Partial derivatives for the
+ * first-passage time distribution in Wiener diffusion models.
+ * *Journal of Mathematical Psychology, 103*, 102550.
+ * https://doi.org/10.1016/j.jmp.2021.102550
+ * - Navarro, D. J., & Fuss, I. G. (2009). Fast and accurate calculations for
+ * first-passage times in Wiener diffusion models.
+ * *Journal of Mathematical Psychology, 53*(4), 222–230.
+ * https://doi.org/10.1016/j.jmp.2009.02.003
+ */
+
+template <bool propto, typename T_y, typename T_alpha, typename T_delta,
+          typename T_beta, typename T_t0, typename T_sv, typename T_sw,
+          typename T_st0>
+inline return_type_t<T_y, T_alpha, T_delta, T_beta, T_t0, T_sv, T_sw, T_st0>
+wiener_full_lpdf(const T_y& y, const T_alpha& a, const T_delta& v,
+                 const T_beta& w, const T_t0& t0, const T_sv& sv,
+                 const T_sw& sw, const T_st0& st0) {
+  double precision = 1e-4;
+
+  return internal::wiener_full_prec_impl_lpdf<propto, T_y, T_alpha, T_delta,
+                                              T_beta, T_t0, T_sv, T_sw, T_st0>(
+      "wiener_full_lpdf", y, a, v, w, t0, sv, sw, st0, precision);
+}
```

Comment:
You don't need separate definitions here for `wiener_full` and `wiener_full_prec`. Since the precision is the last argument, you can just specify it as a default argument which the user can replace if they want:

```suggestion
wiener_full_lpdf(const T_y& y, const T_alpha& a, const T_delta& v,
                 const T_beta& w, const T_t0& t0, const T_sv& sv,
                 const T_sw& sw, const T_st0& st0, const double precision = 1e-4) {
  return internal::wiener_full_prec_impl_lpdf<propto, T_y, T_alpha, T_delta,
                                              T_beta, T_t0, T_sv, T_sw, T_st0>(
      "wiener_full_lpdf", y, a, v, w, t0, sv, sw, st0, precision);
}
```

Reply:
One question here:
How does it work in detail? I tried to do the following:
1) I changed the function as you suggested.
2) I changed the stan model file such that it calls wiener_full_lpdf a) with and b) without a precision value
3) I added the following line in the Stan_math_signatures.ml in stanc3:
`; ([Lpdf], "wiener_full", [DReal; DReal; DReal; DReal; DReal; DReal; DReal; DReal; DReal], SoA) `
Then, the function call with precision works. But the function call with no precision provided does not work. When I change this line and delete the last entry, such that the signature accepts only 8 arguments instead of 9:
`; ([Lpdf], "wiener_full", [DReal; DReal; DReal; DReal; DReal; DReal; DReal; DReal], SoA) `
then it is the other way round: The function call with precision does not work anymore but instead the function call without precision provided.
4) I tried to define both signatures in Stan_math_signatures.ml, but then there occurs a 'duplicate key' error in compilation
(I always try to compile from a linux terminal in the cmdstan-folder with `make ... /wiener_full_lpdf_example`)

So, is there a way to define the signature with a varying number of input arguments in the Stan_math_signatures.ml file?

Reply:
@WardBrian would you be able to clarify the syntax for specifying a signature with a default argument in Stanc3?

Reply:
I don't think we currently have any distributions with this behavior, so I wouldn't expect it to work out of the box, but it's definitely possible. 

Listing both of them is the correct thing to do, but it causes an error in some of our deprecation detection code which assumes the map has unique keys. We can fix this by just discarding one, since all we care about in that code is the name.

Reply:
I believe the 1D integrator has an optional argument: https://mc-stan.org/docs/functions-reference/functions-1d-integrator.html#call-to-the-1d-integrator

But I'm not sure whether that's because it has some special case backend in stanc3 or similar

Reply:
I'm sorry @WardBrian , I think didn't get this.  Do I have to merge the develop branch into my branch and then it will work the way above? Then I could change it when @andrjohns made his changes.

Reply:
I would worry about the stanc3 PR after this is merged. We can definitely have it such that the distribution has optional arguments, it just will require a small amount of tweaking to some existing code which assumes this isn't the case.